Videoplayer code cleanup & race condition fix

MaxKellermann opened the pull request.

garbear commented on xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitleStream.cpp:
```diff
@@ -73,7 +72,7 @@ bool CDVDSubtitleStream::Open(const std::string& strFile)
     } while (read > 0);
 
-    delete pInputStream;
+    pInputStream.reset();
```
returning should automatically reset, correct? this line can be dropped

MaxKellermann replied:
On 2016/03/24 21:33, Garrett Brown <notifications@github.com> wrote:
> > @@ -73,7 +72,7 @@ bool CDVDSubtitleStream::Open(const std::string& strFile)
> >          totalread += read;
> >      } while (read > 0);
> >  
> > -    delete pInputStream;
> > +    pInputStream.reset();
> 
> returning should automatically reset, correct? this line can be dropped

Yes, but the old code did it here, so I did the same, to retain old
behavior.  Sometimes, there can be an advantage in freeing an object
earlier in the scope, and in this case, I wasn't sure if that early
delete was intentional.  Of course, I'll drop it if you prefer so.

garbear commented on xbmc/cores/VideoPlayer/DVDDemuxSPU.cpp:
```diff
@@ -365,7 +365,7 @@ CDVDOverlaySpu* CDVDDemuxSPU::ParseRLE(CDVDOverlaySpu* pSPU, uint8_t* pUnparsedD
 
   // allocate a buffer for the result
-  unsigned __int16* p_dest = (unsigned __int16*)pSPU->result;
+  uint16_t* p_dest = (uint16_t*)pSPU->result;
```
mind changing to reinterpret_cast while you're at it?

MaxKellermann replied:
On 2016/03/24 21:34, Garrett Brown <notifications@github.com> wrote:
> > @@ -365,7 +365,7 @@ CDVDOverlaySpu* CDVDDemuxSPU::ParseRLE(CDVDOverlaySpu* pSPU, uint8_t* pUnparsedD
> >    unsigned int i_x, i_y;
> >  
> >    // allocate a buffer for the result
> > -  unsigned __int16* p_dest = (unsigned __int16*)pSPU->result;
> > +  uint16_t* p_dest = (uint16_t*)pSPU->result;
> 
> mind changing to reinterpret_cast while you're at it?

Yes, but IMHO such things have to be done in a different commit.  It
would be confusing to do two things at once, it would be more
difficult to review.

garbear commented on xbmc/cores/VideoPlayer/VideoPlayer.cpp:
```diff
@@ -890,16 +890,17 @@ void CVideoPlayer::OpenDefaultStreams(bool reset)
+  for (const auto &stream : m_SelectionStreams.Get(STREAM_VIDEO, PredicateVideoPriority))
   {
-    if(OpenStream(m_CurrentVideo, it->demuxerId, it->id, it->source, reset))
+    if(OpenStream(m_CurrentVideo, stream.demuxerId, stream.id, stream.source, reset)) {
```
bracket on second line to match our code style

garbear commented on xbmc/cores/VideoPlayer/VideoPlayer.cpp:
```diff
@@ -90,16 +90,14 @@ using namespace KODI::MESSAGING;
+                   [type, source](const SelectionStream &stream){
+                     return (type == STREAM_NONE || stream.type == type) &&
+                       (source == 0 || stream.source == source);
+                   });
```
This can stay as is, but I think that readability is improved if newlines are single-indented by structuring these awkward lambas like this:

```c++
auto new_end = std::remove_if(m_Streams.begin(), m_Streams.end(),
  [type, source](const SelectionStream &stream)
  {
    return (type == STREAM_NONE || stream.type == type) &&
           (source == 0 || stream.source == source);
  });
```
other examples from a PR of mine: https://github.com/xbmc/xbmc/blob/master/xbmc/input/joysticks/generic/ButtonMapping.cpp#L141

FernetMenta replied:
this entire class needs a re-write

garbear commented:
Thanks for the cleanup. Code improvements are nice, but IMO what's more important is that it deepens your understanding of what the hell's going on in some of Kodi's larger and more complex areas :) Keep up the good work.

FernetMenta commented on xbmc/cores/VideoPlayer/VideoPlayer.cpp:
```diff
@@ -117,6 +117,7 @@ SelectionStream& CSelectionStreams::Get(StreamType type, int index)
 
 std::vector<SelectionStream> CSelectionStreams::Get(StreamType type)
 {
+  CSingleLock lock(m_section);
```
the actual problem is this this is not supposed to be called from other threads

MaxKellermann replied:
What does that mean? What is "other thread"? Does that mean the class doesn't need protection at all, and the mutex should be removed? There is protection currently, and it's incorrect. Either remove it completely, or fix it (like I did).

FernetMenta replied:
VideoPlayer has its own thread and only this thread is supposed to access this method. It was designed only for use in OpenDefaultStreams: https://github.com/xbmc/xbmc/pull/700/commits/240692b2a3a4b9675bb4c972e745d7a1e8e1c3de

Unfortunately folks started to use it for different purpose like exposing information to the GUI. When calling this method the vector is copied twice which is a performance issue.

MaxKellermann replied:
On 2016/03/25 08:26, Rainer Hochecker <notifications@github.com> wrote:
> VideoPlayer has its own thread and only this thread is supposed to access this method. It was designed only for use in OpenDefaultStreams: https://github.com/xbmc/xbmc/pull/700/commits/240692b2a3a4b9675bb4c972e745d7a1e8e1c3de

I see.  So that's the current state, which we must deal with.  Which
means my commit is necessary.

If you don't want my commit, plan B is to remove all uses of
CSelectionStreams from other parts of Kodi, then document
CSelectionStreams to be **not** thread-safe, and remove locking
overhead.

What do you want me to do?

> Unfortunately folks started to use it for different purpose like exposing information to the GUI. When calling this method the vector is copied twice which is a performance issue.

Twice?  It is copied once (i.e. data from "m_Streams" is copied to
local variable "streams").  The "return streams" is not a copy,
it's "moved" to the caller.  That is just a pointer exchange.  Some
C++ compilers might even be able to eliminate this move (maybe
gcc+clang do this already, I'm not sure).

FernetMenta replied:
> If you don't want my commit, plan B is to remove all uses of

Nope. Only this particular method does not require locking. See your commit message. It's only VideoPlayer thread that can change the vector. The short term solution is not calling this method from outside.

MaxKellermann replied:
So instead of the race condition fix, will you accept a patch that adds API documentation stating that using this method makes an instances thread-unsafe?

FernetMenta replied:
Usage of this method from any other thread needs to be removed. See how it is used. The streams are copied to the local variable and then a second time (subs in this case).

```
bool CVideoPlayer::GetStreamDetails(CStreamDetails &details)
{
  if (m_pDemuxer)
  {
    std::vector<SelectionStream> subs = m_SelectionStreams.Get(STREAM_SUBTITLE);
```

This snippet shows the next no-go. The outside world must not access demuxer

MaxKellermann replied:
I understand, the only thing I don't understand is what you want me to do. I made three suggestions, and you didn't agree to any of these. I'd like to post something that meets your agreement, but it appears impossible.

FernetMenta replied:
>  what you want me to do

remove the calls to m_SelectionStreams.Get in methods called from the outside.

MaxKellermann replied:
Please tell me where class CSelectionStreams is accessed from outside of the player thread. I found nothing. Or more precise: it is used in the CVideoPlayer constructor, at a time where the player thread has not yet been spawned. That is safe.

Am I right to assume that locking in CSelectionStreams is completely unnecessary?

FernetMenta replied:
> Please tell me where class CSelectionStreams is accessed from outside of the player thread

see the snippet in my post above: i.e.
CVideoPlayer::GetStreamDetails or
CVideoPlayer::GetAudioStreamInfo
(there is more)

> Am I right to assume that locking in CSelectionStreams is completely unnecessary?

If a user wants to change audio or subtitle track the UI needs to have access to the currently available streams. The UI also accesses this object for getting info to display on the OSD.
Having the GUI poll player data frequently is not ideal. It is much better to provide an extra object for the UI and player updates this object after something has changed.
That's the vision and we don't need to get to final in a single iteration.

MaxKellermann replied:
On 2016/03/25 19:08, Rainer Hochecker <notifications@github.com> wrote:
> see the snippet in my post above: i.e.
> CVideoPlayer::GetStreamDetails or
> CVideoPlayer::GetAudioStreamInfo
> (there is more)

OK, so these methods are called from outside the player thread.  I
thought what you meant was users of CSelectionStreams from classes
other than VideoPlayer.  I see.

I will check these methods and see what I can do to implement lockless
read-only access.  (Out of this PR's scope.)

FernetMenta commented:
please submit in smaller junks. VideoPlayer is in the process of transformation. There are lots of areas under construction and doing cosmetics to this code is just a change without a step forward. 
There are many changes I agree with but with others I don't.

MaxKellermann commented:
On 2016/03/24 22:43, Rainer Hochecker <notifications@github.com> wrote:
> please submit in smaller junks. VideoPlayer is in the process of transformation. There are lots of areas under construction and doing cosmetics to this code is just a change without a step forward. 
> There are many changes I agree with but with others I don't.

Splitting to smaller junks is difficult without knowing which ones you
agree with.  Can you just cherry-pick the ones you agree with, and
then we'll talk about the rest?

FernetMenta commented:
ok

MaxKellermann commented:
Rebased with the following changes:
- indentation of lambdas changed as suggested
- explicit release() call removed, even though I think it was better before

MaxKellermann commented:
Rebased and dropped the patch "VideoPlayer: fix race condition in Get(StreamType)", because @FernetMenta asked me to rewrite it. I believe that all of the locking was never necessary (see https://github.com/xbmc/xbmc/pull/9442#discussion_r57465958), and thus a reconsideration of the class is necessary.

As soon as we have a result, I'll implement the solution. Until then, please merge the remaining commits.

MaxKellermann commented:
Rebase & changed indent according to https://github.com/xbmc/xbmc/pull/9442#discussion-diff-57383380

wsnipex commented:
jenkins build this please

FernetMenta commented:
we agreed that I cherry-pick from this PR. no need to put load on Jenkins here.

MaxKellermann commented:
Rebased, dropping the patches that were merged by cherry-picking PR https://github.com/xbmc/xbmc/pull/9454

Darwin/Android builds probably still fail due to their broken system headers. To be checked.

MaxKellermann commented:
So, this turns out not to be broken system headers, but a kludge in utils/Regexp.h which looks so ugly, it's surprising that it breaks only Android and Darwin. The offending code (abbreviated):

    namespace PCRE {
    #include <pcre.h>
    }

This tries to be clever by moving all `pcre.h` definitions into a separate namespace. The problem is that `pcre.h` includes `stdlib.h`, thus all if `stdlib.h`'s definitions are now also living in the `PCRE` namespace. This obviously breaks anything that searches for `stdlib.h` definitions in the root namespace.

Ouch!

So this code has always been there, added by the extraordinarily huge (and undocumented) commit https://github.com/xbmc/xbmc/commit/45285e8

MartijnKaijser commented:
@MaxKellermann yes that commit is some breakpoint in time :) Should you wish to dig deeper the continued history is here https://github.com/xbmc/xbmc-antiquated

garbear commented:
I've run into this problem before. I solved it by creating a wrapper header, e.g. `pcre_wrapper.h`, that included pcre.h's headers outside the namespace, then included pcre inside the namespace. e.g.

```c++
// Include all headers used by pcre.h
#include <stdlib.h>
namespace PCRE
{
#include <pcre.h>
}
```

MaxKellermann commented:
@garbear, had the same idea to work around this for now. Though I'd like to get rid of this fragile namespace hack. What purpose does it serve?

garbear commented:
It's plausible that pcre.h contains one or more symbols that conflict with Kodi in the global namespace

MaxKellermann commented:
I submitted https://github.com/xbmc/xbmc/pull/9494 to fast-track the workaround and the include cleanup. I think the rest was not-yet-accepted by @FernetMenta, so after merging 9494, we should discuss the rest of this PR.

MaxKellermann commented:
After #9494 has been merged, I've rebased this branch. Please discuss. (@FernetMenta?)

FernetMenta commented:
@MaxKellermann I'll look at it tomorrow. Need to fix some bugs first.

MaxKellermann committed:
VideoPlayer: use std::remove_if() for efficient removal

Changes quadratic behavior in Clear() to linear.

MaxKellermann committed:
VideoPlayer/DVDDemuxSPU: make destructor non-virtual

Nobody derives from this class.

MaxKellermann committed:
VideoPlayer: use std::copy_if() in Get(StreamType)

Reduce overhead.

MaxKellermann commented:
After merging #9530, this is a rebase with 4 remaining unmerged commits.

FernetMenta commented:
Adding overrides is correct but this interface needs cleanup first. Means looking at the methods one by one and decide if those still make sense. In order to do so history and git blame are valuable tools. https://github.com/xbmc/xbmc/pull/9442/commits/9b1032926ff799aa826cfd88bdb0e0a9d93c12af would make this work harder. This is why I don't pick it in this stage.

ee89f85c434f8f230d392860f32d96ac79d2ff49 why don't you make the class final? That would make it obvious that it is not supposed to inherit from.

I am ok with the other two in selection streams if @ace20022 is as well.

MaxKellermann commented:
Having the "overrides" makes it easier to change the API, because you'll get compiler errors if you remove a `virtual` from the base class without removing them from all derived classes.

I'll add `final` soon. I wasn't sure if MSVC knows `final` - I've become cautious with C++11 keywords because your MSVC version is too old for some C++11 features.

FernetMenta commented:
I won't take the overrides commit for the reasons given. Feel free to review individual methods and submit PRs, but I won't accept this bulk operation.

afedchin commented:
@MaxKellermann don't worry about vs, it knows about final

MaxKellermann committed:
VideoPlayer/DVDDemuxSPU: make the class final

MaxKellermann commented:
Dropped the `override` commit from this PR, and added `final` commit.

FernetMenta commented:
thanks
jenkins build this please

ace20022 commented:
> I am ok with the other two in selection streams if @ace20022 is as well.

 I have no objections.

FernetMenta commented:
Windows fails with:

```
7>..\..\xbmc\cores\VideoPlayer\VideoPlayer.cpp(120): error C2039: 'back_inserter': is not a member of 'std' [c:\jenkins\slave\workspace\WIN-32\project\VS2010Express\XBMC.vcxproj]
```

ace20022 commented:
> Windows fails with:

```#include <iterator>``` is missing

MaxKellermann commented:
Oops. Should be fixed now. Please try again.

FernetMenta commented:
jenkins build this please

FernetMenta commented:
thanks again!
